File: frontend/src/types/insp/indicator.ts
import type { LongId } from '@/types/common'

/**
 * 检查平台 - 评价指标类型定义
 *
 * Phase 4 评级引擎完美架构 (2026-05-23) 新增字段:
 *   sourceSectionIds[] / triggerMode / countThreshold / weightsBySection /
 *   rankDirection / missingPolicy(枚举化) / latePolicy / submissionDateField
 * 评级结果聚合根 IndicatorResult (DRAFT/PUBLISHED/SUPERSEDED 三态机) 类型也在此.
 */

export type TriggerMode = 'TIME_WINDOW' | 'COUNT' | 'MANUAL'
export type RankDirection = 'ASC' | 'DESC' | 'NONE'
export type MissingPolicyEnum = 'IGNORE' | 'ZERO' | 'MAX' | 'WAIT'
export type LatePolicy = 'REVISE_ORIGINAL' | 'CARRY_FORWARD' | 'EXCLUDE'
export type SubmissionDateField = 'TASK_DATE' | 'COMPLETED_AT'
export type ResultStatus = 'DRAFT' | 'PUBLISHED' | 'SUPERSEDED'

export interface Indicator {
  id: LongId
  tenantId: LongId
  projectId: LongId
  parentIndicatorId: LongId | null
  name: string
  indicatorType: 'LEAF' | 'COMPOSITE'
  // LEAF fields — 旧单值 (向后兼容, sourceSectionIds 优先)
  sourceSectionId: LongId | null
  /** Phase 4 多分区: 一个 LEAF 可跨多个分区. */
  sourceSectionIds?: LongId[]
  sourceAggregation: string | null
  // COMPOSITE fields
  compositeAggregation: string | null
  missingPolicy: string | null
  // Normalization
  normalization: string | null
  normalizationConfig: string | null
  // Common
  evaluationPeriod: string
  gradeSchemeId: LongId | null
  evaluationMethod: string | null
  gradeThresholds: string | null
  sortOrder: number
  // ── Phase 4 评级引擎新字段 ──
  triggerMode?: TriggerMode | null
  countThreshold?: number | null
  weightsBySection?: Record<string, number> | null
  rankDirection?: RankDirection | null
  latePolicy?: LatePolicy | null
  submissionDateField?: SubmissionDateField | null
  createdAt: string
  updatedAt: string | null
  // Client-side tree
  children?: Indicator[]
}

export interface IndicatorScore {
  id: LongId
  indicatorId: LongId
  targetId: LongId
  targetName: string | null
  targetType: string | null
  periodStart: string
  periodEnd: string
  score: number | null
  gradeCode: string | null
  gradeName: string | null
  gradeColor: string | null
  sourceCount: number
  detail: string | null
}

/** Phase 4 评级结果聚合根 — 版本化快照 + 三态机. */
export interface IndicatorResult {
  id: LongId
  tenantId: LongId
  indicatorId: LongId
  targetId: LongId
  targetName?: string | null
  targetType?: string | null
  periodKey: string
  periodStart?: string | null
  periodEnd?: string | null
  /** 同 (indicator,target,period) 修订版本号 — 0=首版, 后续递增. */
  revision: number
  status: ResultStatus
  value: number | null
  rankPosition: number | null
  totalRanked: number | null
  gradeCode?: string | null
  gradeName?: string | null
  gradeColor?: string | null
  sourceCount?: number | null
  /** 该版本的算分上下文快照 (JSON). */
  snapshot?: string | null
  /** 替代前一版本时, 该字段记录前一 result id; 首版为 null. */
  supersedesId?: LongId | null
  computedAt: string
  publishedAt?: string | null
  createdAt?: string
  updatedAt?: string | null
}

export interface CreateLeafIndicatorRequest {
  projectId: LongId
  parentIndicatorId: LongId | null
  name: string
  /** 单值 (向后兼容). 优先用 sourceSectionIds. */
  sourceSectionId?: LongId | null
  /** Phase 4 多分区 — 至少 1 个. */
  sourceSectionIds?: LongId[]
  sourceAggregation: string
  evaluationPeriod: string
  // ── Phase 4 评级引擎新字段 ──
  triggerMode?: TriggerMode
  countThreshold?: number
  weightsBySection?: Record<string, number>
  rankDirection?: RankDirection
  missingPolicy?: MissingPolicyEnum
  latePolicy?: LatePolicy
  submissionDateField?: SubmissionDateField
  // 等级 / 归一
  gradeSchemeId?: LongId | null
  normalization?: string
  normalizationConfig?: string
  evaluationMethod?: string
  gradeThresholds?: string
  sortOrder?: number
}

export interface CreateCompositeIndicatorRequest {
  projectId: LongId
  parentIndicatorId: LongId | null
  name: string
  compositeAggregation: string
  missingPolicy: string
  evaluationPeriod: string
  sourceSectionId?: LongId
  gradeSchemeId?: LongId | null
  normalization?: string
  normalizationConfig?: string
  evaluationMethod?: string
  gradeThresholds?: string
  sortOrder?: number
}

export interface UpdateIndicatorRequest {
  name?: string
  evaluationPeriod?: string
  gradeSchemeId?: LongId | null
  sourceSectionId?: LongId | null
  /** Phase 4 多分区. */
  sourceSectionIds?: LongId[]
  sourceAggregation?: string
  compositeAggregation?: string
  missingPolicy?: string
  normalization?: string
  normalizationConfig?: string
  evaluationMethod?: string
  gradeThresholds?: string
  sortOrder?: number
  // ── Phase 4 评级引擎新字段 ──
  triggerMode?: TriggerMode
  countThreshold?: number
  weightsBySection?: Record<string, number>
  rankDirection?: RankDirection
  latePolicy?: LatePolicy
  submissionDateField?: SubmissionDateField
}

export interface ManualEvaluateRequest {
  indicatorId: LongId
  /** ISO date YYYY-MM-DD. */
  startDate: string
  endDate: string
}

export const SOURCE_AGG_OPTIONS = [
  { value: 'AVG', label: '取平均' },
  { value: 'MAX', label: '取最高' },
  { value: 'MIN', label: '取最低' },
  { value: 'LATEST', label: '取最新' },
  { value: 'SUM', label: '求和' },
]

export const COMPOSITE_AGG_OPTIONS = [
  { value: 'WEIGHTED_AVG', label: '加权平均' },
  { value: 'SUM', label: '求和' },
  { value: 'AVG', label: '平均' },
  { value: 'MAX', label: '取最高' },
  { value: 'MIN', label: '取最低' },
]

export const MISSING_POLICY_OPTIONS = [
  { value: 'SKIP', label: '跳过（重新分配权重）' },
  { value: 'CARRY_FORWARD', label: '沿用上期' },
  { value: 'MARK_INCOMPLETE', label: '标记不完整' },
]

export const EVAL_PERIOD_OPTIONS = [
  { value: 'PER_TASK', label: '每次检查' },
  { value: 'DAILY', label: '每天' },
  { value: 'WEEKLY', label: '每周' },
  { value: 'MONTHLY', label: '每月' },
]

export const EVALUATION_METHOD_OPTIONS = [
  { value: 'SCORE_RANGE', label: '绝对分数', description: '按原始分数划分等级' },
  { value: 'PERCENT_RANGE', label: '得分率', description: '按得分百分比划分（适配不同满分）' },
  { value: 'RANK_COUNT', label: '排名前N', description: '按排名位次划分（如前3名）' },
  { value: 'RANK_PERCENT', label: '排名百分比', description: '按排名百分位划分（如前20%）' },
]

export const NORMALIZATION_OPTIONS = [
  { value: 'NONE', label: '不归一化', description: '使用原始分数直接比较', icon: 'equal' },
  { value: 'RELATION_COUNT', label: '按关联数量', description: '除以目标关联的实体数量（如成员人数）', icon: 'users' },
  { value: 'FIXED_VALUE', label: '固定除数', description: '除以一个固定的数值', icon: 'hash' },
  { value: 'PERCENTAGE', label: '转百分比', description: '转换为百分比（分数÷满分×100）', icon: 'percent' },
]

// ==================== Phase 4 新枚举选项 ====================

export const TRIGGER_MODE_OPTIONS: { value: TriggerMode; label: string; description: string }[] = [
  { value: 'TIME_WINDOW', label: '时间窗口', description: '按 评估周期 自动滚动评估 (每天/每周/每月)' },
  { value: 'COUNT', label: '次数', description: '同目标累计达到 N 次提交后触发一次评估' },
  { value: 'MANUAL', label: '手动', description: '管理员在 评级结果 页主动触发' },
]

export const RANK_DIRECTION_OPTIONS: { value: RankDirection; label: string }[] = [
  { value: 'DESC', label: '越大越好 (高分在前)' },
  { value: 'ASC', label: '越小越好 (低分在前)' },
  { value: 'NONE', label: '不排名' },
]

export const MISSING_POLICY_ENUM_OPTIONS: { value: MissingPolicyEnum; label: string; description: string }[] = [
  { value: 'IGNORE', label: '忽略', description: '该目标本期无数据则跳过, 不生成结果' },
  { value: 'ZERO', label: '记零分', description: '无数据时按 0 分入榜' },
  { value: 'MAX', label: '记满分', description: '无数据时按上限计入 (适合"未违规即满分"场景)' },
  { value: 'WAIT', label: '等待', description: '本期不评, 等下次数据补齐再发布' },
]

export const LATE_POLICY_OPTIONS: { value: LatePolicy; label: string; description: string }[] = [
  { value: 'REVISE_ORIGINAL', label: '修订原版', description: '迟到提交触发当期重算, 旧版置 SUPERSEDED' },
  { value: 'CARRY_FORWARD', label: '顺延下期', description: '迟到数据并入下一评估期' },
  { value: 'EXCLUDE', label: '排除', description: '迟到提交永久不进入评级' },
]

export const SUBMISSION_DATE_FIELD_OPTIONS: { value: SubmissionDateField; label: string }[] = [
  { value: 'TASK_DATE', label: '按任务日期' },
  { value: 'COMPLETED_AT', label: '按完成时间' },
]

export const RESULT_STATUS_OPTIONS: { value: ResultStatus; label: string; tone: string }[] = [
  { value: 'DRAFT', label: '草稿', tone: 'warn' },
  { value: 'PUBLISHED', label: '已发布', tone: 'pass' },
  { value: 'SUPERSEDED', label: '已被替代', tone: 'pending' },
]

File: backend/src/main/java/com/school/management/domain/inspection/model/v7/scoring/Indicator.java
package com.school.management.domain.inspection.model.v7.scoring;

import com.school.management.domain.shared.Entity;

import java.time.LocalDateTime;

/**
 * V7 评价指标 — 聚合根
 * <p>
 * 指标树节点，分 LEAF（叶子，绑定数据来源分区）和 COMPOSITE（复合，聚合子指标）两种类型。
 */
public class Indicator implements Entity<Long> {

    private Long id;
    private Long tenantId;
    private Long projectId;
    private Long parentIndicatorId;
    private String name;
    private String indicatorType; // LEAF | COMPOSITE

    // LEAF fields
    private Long sourceSectionId;
    private String sourceAggregation; // AVG | MAX | MIN | LATEST | SUM

    // COMPOSITE fields
    private String compositeAggregation; // WEIGHTED_AVG | SUM | AVG | MIN | MAX
    private String missingPolicy; // SKIP | CARRY_FORWARD | MARK_INCOMPLETE

    // Normalization
    private String normalization;       // NONE | RELATION_COUNT | FIXED_VALUE | PERCENTAGE
    private String normalizationConfig; // JSON config

    private String evaluationPeriod; // PER_TASK | DAILY | WEEKLY | MONTHLY
    private Long gradeSchemeId;
    private String evaluationMethod;    // SCORE_RANGE | PERCENT_RANGE | RANK_COUNT | RANK_PERCENT
    private String gradeThresholds;     // JSON: [{"gradeCode":"RED","value":90},...]
    private Integer sortOrder;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    protected Indicator() {
    }

    private Indicator(Builder builder) {
        this.id = builder.id;
        this.tenantId = builder.tenantId;
        this.projectId = builder.projectId;
        this.parentIndicatorId = builder.parentIndicatorId;
        this.name = builder.name;
        this.indicatorType = builder.indicatorType;
        this.sourceSectionId = builder.sourceSectionId;
        this.sourceAggregation = builder.sourceAggregation;
        this.compositeAggregation = builder.compositeAggregation;
        this.missingPolicy = builder.missingPolicy;
        this.normalization = builder.normalization != null ? builder.normalization : "NONE";
        this.normalizationConfig = builder.normalizationConfig;
        this.evaluationPeriod = builder.evaluationPeriod != null ? builder.evaluationPeriod : "PER_TASK";
        this.gradeSchemeId = builder.gradeSchemeId;
        this.evaluationMethod = builder.evaluationMethod != null ? builder.evaluationMethod : "PERCENT_RANGE";
        this.gradeThresholds = builder.gradeThresholds;
        this.sortOrder = builder.sortOrder != null ? builder.sortOrder : 0;
        this.createdAt = builder.createdAt != null ? builder.createdAt : LocalDateTime.now();
        this.updatedAt = builder.updatedAt;
    }

    // ── Factory ──────────────────────────────────────────────

    public static Indicator createLeaf(Long projectId, String name,
                                       Long sourceSectionId, String sourceAggregation,
                                       String evaluationPeriod, Long gradeSchemeId) {
        Indicator ind = new Indicator();
        ind.projectId = projectId;
        ind.name = name;
        ind.indicatorType = "LEAF";
        ind.sourceSectionId = sourceSectionId;
        ind.sourceAggregation = sourceAggregation != null ? sourceAggregation : "AVG";
        ind.evaluationPeriod = evaluationPeriod != null ? evaluationPeriod : "PER_TASK";
        ind.gradeSchemeId = gradeSchemeId;
        ind.sortOrder = 0;
        ind.createdAt = LocalDateTime.now();
        return ind;
    }

    public static Indicator createComposite(Long projectId, String name,
                                            String compositeAggregation, String missingPolicy,
                                            String evaluationPeriod, Long gradeSchemeId) {
        Indicator ind = new Indicator();
        ind.projectId = projectId;
        ind.name = name;
        ind.indicatorType = "COMPOSITE";
        ind.compositeAggregation = compositeAggregation != null ? compositeAggregation : "AVG";
        ind.missingPolicy = missingPolicy != null ? missingPolicy : "SKIP";
        ind.evaluationPeriod = evaluationPeriod != null ? evaluationPeriod : "WEEKLY";
        ind.gradeSchemeId = gradeSchemeId;
        ind.sortOrder = 0;
        ind.createdAt = LocalDateTime.now();
        return ind;
    }

    public static Indicator reconstruct(Builder builder) {
        return new Indicator(builder);
    }

    // ── Commands ─────────────────────────────────────────────

    public void update(String name, String evaluationPeriod,
                       Long gradeSchemeId, Long sourceSectionId, String sourceAggregation,
                       String compositeAggregation, String missingPolicy,
                       String normalization, String normalizationConfig,
                       String evaluationMethod, String gradeThresholds,
                       Integer sortOrder) {
        this.name = name;
        this.evaluationPeriod = evaluationPeriod;
        this.gradeSchemeId = gradeSchemeId;
        this.sourceSectionId = sourceSectionId;
        this.sourceAggregation = sourceAggregation;
        this.compositeAggregation = compositeAggregation;
        this.missingPolicy = missingPolicy;
        this.normalization = normalization;
        this.normalizationConfig = normalizationConfig;
        this.evaluationMethod = evaluationMethod;
        this.gradeThresholds = gradeThresholds;
        if (sortOrder != null) {
            this.sortOrder = sortOrder;
        }
        this.updatedAt = LocalDateTime.now();
    }

    public void setParentIndicatorId(Long parentIndicatorId) {
        this.parentIndicatorId = parentIndicatorId;
    }

    public void setTenantId(Long tenantId) {
        this.tenantId = tenantId;
    }

    public void setSortOrder(Integer sortOrder) {
        this.sortOrder = sortOrder;
        this.updatedAt = LocalDateTime.now();
    }

    // ── Queries ──────────────────────────────────────────────

    public boolean isLeaf() {
        return "LEAF".equals(indicatorType);
    }

    public boolean isComposite() {
        return "COMPOSITE".equals(indicatorType);
    }

    public boolean isRoot() {
        return parentIndicatorId == null;
    }

    // ── Getters ──────────────────────────────────────────────

    @Override
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Long getTenantId() {
        return tenantId;
    }

    public Long getProjectId() {
        return projectId;
    }

    public Long getParentIndicatorId() {
        return parentIndicatorId;
    }

    public String getName() {
        return name;
    }

    public String getIndicatorType() {
        return indicatorType;
    }

    public Long getSourceSectionId() {
        return sourceSectionId;
    }

    public String getSourceAggregation() {
        return sourceAggregation;
    }

    public String getCompositeAggregation() {
        return compositeAggregation;
    }

    public String getMissingPolicy() {
        return missingPolicy;
    }

    public String getNormalization() { return normalization; }
    public String getNormalizationConfig() { return normalizationConfig; }

    public String getEvaluationPeriod() {
        return evaluationPeriod;
    }

    public Long getGradeSchemeId() {
        return gradeSchemeId;
    }

    public String getEvaluationMethod() {
        return evaluationMethod;
    }

    public String getGradeThresholds() {
        return gradeThresholds;
    }

    public Integer getSortOrder() {
        return sortOrder;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    // ── Builder ──────────────────────────────────────────────

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private Long id;
        private Long tenantId;
        private Long projectId;
        private Long parentIndicatorId;
        private String name;
        private String indicatorType;
        private Long sourceSectionId;
        private String sourceAggregation;
        private String compositeAggregation;
        private String missingPolicy;
        private String normalization;
        private String normalizationConfig;
        private String evaluationPeriod;
        private Long gradeSchemeId;
        private String evaluationMethod;
        private String gradeThresholds;
        private Integer sortOrder;
        private LocalDateTime createdAt;
        private LocalDateTime updatedAt;

        public Builder id(Long id) { this.id = id; return this; }
        public Builder tenantId(Long tenantId) { this.tenantId = tenantId; return this; }
        public Builder projectId(Long projectId) { this.projectId = projectId; return this; }
        public Builder parentIndicatorId(Long parentIndicatorId) { this.parentIndicatorId = parentIndicatorId; return this; }
        public Builder name(String name) { this.name = name; return this; }
        public Builder indicatorType(String indicatorType) { this.indicatorType = indicatorType; return this; }
        public Builder sourceSectionId(Long sourceSectionId) { this.sourceSectionId = sourceSectionId; return this; }
        public Builder sourceAggregation(String sourceAggregation) { this.sourceAggregation = sourceAggregation; return this; }
        public Builder compositeAggregation(String compositeAggregation) { this.compositeAggregation = compositeAggregation; return this; }
        public Builder missingPolicy(String missingPolicy) { this.missingPolicy = missingPolicy; return this; }
        public Builder normalization(String normalization) { this.normalization = normalization; return this; }
        public Builder normalizationConfig(String normalizationConfig) { this.normalizationConfig = normalizationConfig; return this; }
        public Builder evaluationPeriod(String evaluationPeriod) { this.evaluationPeriod = evaluationPeriod; return this; }
        public Builder gradeSchemeId(Long gradeSchemeId) { this.gradeSchemeId = gradeSchemeId; return this; }
        public Builder evaluationMethod(String evaluationMethod) { this.evaluationMethod = evaluationMethod; return this; }
        public Builder gradeThresholds(String gradeThresholds) { this.gradeThresholds = gradeThresholds; return this; }
        public Builder sortOrder(Integer sortOrder) { this.sortOrder = sortOrder; return this; }
        public Builder createdAt(LocalDateTime createdAt) { this.createdAt = createdAt; return this; }
        public Builder updatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; return this; }

        public Indicator build() {
            return new Indicator(this);
        }
    }
}
